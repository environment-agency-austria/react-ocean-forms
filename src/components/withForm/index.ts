export * from './withForm';
export * from './withForm.types';
