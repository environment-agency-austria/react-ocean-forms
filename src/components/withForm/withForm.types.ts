/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @packageDocumentation
 * @module withForm
 */
import { IFormContext } from '../FormContext';

/**
 * Props for consumers of withForm props
 */
export interface IFormContextProps {
  /**
   * Form context of the parent form
   */
  context: IFormContext;
}
