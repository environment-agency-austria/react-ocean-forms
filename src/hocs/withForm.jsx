/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import FormContext from '../FormContext';
import { getDisplayName } from '../utils';

/**
 * High order component for consuming the form context
 */
export default function withForm(Component) {
  /**
   * Component that injects the form context prop
   * to the wrapped component
   */
  function FormComponent(props) {
    return (
      <FormContext.Consumer>
        {context => <Component {...props} context={context} />}
      </FormContext.Consumer>
    );
  }

  FormComponent.displayName = `FormComponent(${getDisplayName(Component)})`;
  return FormComponent;
}
