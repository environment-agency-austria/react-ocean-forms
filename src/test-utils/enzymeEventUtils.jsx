/**
 * Defines some often used keycodes for
 * event testing
 */
export const KEYCODE = {
  BACKSPACE: 8,
  TAB: 9,
  ENTER: 13,
  ESCAPE: 27,
  DOWN_ARROW: 40,
};

/**
 * Generates a mocked event to be used with
 * enzymes simulate in situations where the
 * tested code wants to call preventDefault
 * @param {object} props Additional event props
 */
export const mockEvent = props => ({
  preventDefault: () => {},
  ...props,
});

export default mockEvent;
