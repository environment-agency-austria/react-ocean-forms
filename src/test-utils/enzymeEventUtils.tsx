/**
 * Defines some often used keycodes for
 * event testing
 */
export const KEYCODE = {
  BACKSPACE: 8,
  TAB: 9,
  ENTER: 13,
  ESCAPE: 27,
  DOWN_ARROW: 40,
};

interface IMockEvent {
  preventDefault(): void;
  [key: string]: any;
}

/**
 * Generates a mocked event to be used with
 * enzymes simulate in situations where the
 * tested code wants to call preventDefault
 * @param props Additional event props
 */
export const mockEvent = (props?: any): IMockEvent => ({
  // tslint:disable-next-line:no-empty
  preventDefault: (): void => {},
  ...props,
});

export default mockEvent;
