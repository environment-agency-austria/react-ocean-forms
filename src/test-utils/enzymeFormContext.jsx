import { stringFormatter } from '../utils';

/**
 * Components inside the form module require access to the form context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * user definable form context around them.
 */

/**
 * Creates a form context
 */
export const createMockFormContext = registerCallback => ({
  fieldPrefix: null,

  registerField: jest.fn().mockImplementation((name, state) => registerCallback(name, state)),
  unregisterField: jest.fn(),
  notifyFieldEvent: jest.fn(),

  registerListener: jest.fn().mockImplementation((name, state) => registerCallback(name, state)),
  unregisterListener: jest.fn(),

  getFieldState: jest.fn(),
  getValues: jest.fn(),

  stringFormatter: jest.fn().mockImplementation(stringFormatter),

  busy: false,
  disabled: false,

  asyncValidateOnChange: false,
  asyncValidationWait: 400,
  defaultValues: {},
});

/**
 * Creates a validation object mocking the
 * withValidation hoc
 */
export const createMockValidation = () => ({
  validate: jest.fn(),
  reset: jest.fn(),
  update: jest.fn(),

  isValidating: false,
  valid: true,
  error: null,
});
