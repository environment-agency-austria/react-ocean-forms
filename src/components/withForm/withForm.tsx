/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { getDisplayName, PropsOf, Subtract } from '../../utils';
import { useFormContext } from '../../hooks';
import { IFormContextProps } from './withForm.types';

/**
 * Internal type for form component props. Deals with the
 * defaultProp logic and removes the form context props.
 */
type FormComponentProps<TComp> = Subtract<JSX.LibraryManagedAttributes<TComp, PropsOf<TComp>>, IFormContextProps>;

/**
 * High order component for consuming the form context
 * @deprecated Deprecated in favour of `useFormContext` hook
 */
export const withForm = <TComp extends React.ComponentType<TProps>, TProps extends IFormContextProps = PropsOf<TComp>>(Component: TComp):
React.ComponentType<FormComponentProps<TComp>> => {
  /**
   * We need to cast the component back to a
   * React.ComponentType in order to use it in
   * the FormContext.Consumer. Otherwise
   * typescript will cry about the component
   * not having a constructor / call signatures.
   */
  const CastedComponent = Component as React.ComponentType<TProps>;

  /**
   * Component that injects the form context prop
   * to the wrapped component
   */
  const FormComponent: React.SFC<FormComponentProps<TComp>> = (props: FormComponentProps<TComp>): JSX.Element => {
    const context = useFormContext();

    // @ts-ignore
    return <CastedComponent {...props} context={context} />;
  };
  FormComponent.displayName = `FormComponent(${getDisplayName(Component)})`;

  return FormComponent;
};
