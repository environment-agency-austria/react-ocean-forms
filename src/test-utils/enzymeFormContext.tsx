import { IFieldState, IFormContext, TFormEventListener } from '../components/FormContext';
import { IValidationProp } from '../components/withValidation';
import { stringFormatter } from '../utils';

/**
 * Components inside the form module require access to the form context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * user definable form context around them.
 */

/**
 * Creates a form context
 */
// tslint:disable-next-line:naming-convention
export const createMockFormContext = (registerCallback?: Function): IFormContext => ({
  fieldPrefix: null,

  registerField: jest.fn().mockImplementation((name: string, state: IFieldState): void => {
    if (registerCallback) {
      registerCallback(name, state);
    }
  }),
  unregisterField: jest.fn(),
  notifyFieldEvent: jest.fn(),

  registerListener: jest.fn().mockImplementation((name: string, callback: TFormEventListener): void => {
    if (registerCallback) {
      registerCallback(name, callback);
    }
  }),
  unregisterListener: jest.fn(),

  getFieldState: jest.fn(),
  getValues: jest.fn(),

  stringFormatter: jest.fn().mockImplementation(stringFormatter),
  submit: jest.fn(),

  busy: false,
  disabled: false,
  plaintext: false,

  asyncValidateOnChange: false,
  asyncValidationWait: 400,
  defaultValues: {},
});

/**
 * Creates a validation object mocking the
 * withValidation hoc
 */
// tslint:disable-next-line:naming-convention
export const createMockValidation = (): IValidationProp => ({
  validate: jest.fn(),
  reset: jest.fn(),
  update: jest.fn(),

  isValidating: false,
  isRequired: false,
  valid: true,
  error: null,
});
