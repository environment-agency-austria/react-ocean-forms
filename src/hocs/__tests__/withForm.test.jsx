import React from 'react';
import { shallow } from 'enzyme';

import { createMockFormContext } from '../../test-utils/enzymeFormContext';
import { withForm } from '../withForm';

describe('withForm', () => {
  const formContext = createMockFormContext();
  const TestComponent = () => (<div id="test-component" />);
  const WrappedComponent = withForm(TestComponent);

  const fieldName = 'unitField';

  const setup = props => shallow((
    <WrappedComponent
      name={fieldName}
      context={formContext}
      {...props}
    />
  ));
  const wrapper = setup();

  it('should render without error', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
