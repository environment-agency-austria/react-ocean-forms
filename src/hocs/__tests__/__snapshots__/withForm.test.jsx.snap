// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withForm should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormComponent(TestComponent)
    context={
      Object {
        "asyncValidateOnChange": false,
        "asyncValidationWait": 400,
        "busy": false,
        "defaultValues": Object {},
        "disabled": false,
        "fieldPrefix": null,
        "getFieldState": [MockFunction],
        "getValues": [MockFunction],
        "notifyFieldEvent": [MockFunction],
        "registerField": [MockFunction],
        "registerListener": [MockFunction],
        "stringFormatter": [MockFunction],
        "unregisterField": [MockFunction],
        "unregisterListener": [MockFunction],
      }
    }
    name="unitField"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "Consumer": [Circular],
      "Provider": Object {
        "$$typeof": Symbol(react.provider),
        "_context": [Circular],
      },
      "_calculateChangedBits": null,
      "_changedBits": 0,
      "_changedBits2": 0,
      "_currentRenderer": null,
      "_currentRenderer2": null,
      "_currentValue": "form",
      "_currentValue2": "form",
      "_defaultValue": "form",
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_changedBits2": 0,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": "form",
        "_currentValue2": "form",
        "_defaultValue": "form",
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
